libraries/lib-concurrency/ParallelReduce.h

```diff
@@ -0,0 +1,307 @@
+ object on the leftmost range.
+
+ @tparam Task satisfies the @ref Splittable concept.
+
+ It splitting constructor, `Task(Task &orig, iterator_type iter)` is called zero
+ or more times, only before `run`.
+
+ It also must define:
+ 
+ - `void run(Progress<Increment> &progress)` which typically iterates the
+ range and is called exactly once per task
+ - `void merge(Task &&task, Progress<Increment> &progress)`, which is called
+ exactly once for each task made by the split constructor (absent exceptions),
+ and only after `run` is called on `this` and on `task`
+
+ @tparam Increment defaults to the difference type of the Splittable iterators
+
+ @param task If `(std::distance(task.begin(), task.end()))` is zero or less, no
+ other threads are recruited and execution is serial.
+
+ @param n_threads maximum number of threads to use in the computation, including
+ the calling thread; supply 1 to do all work seriallyx
+
+ @throws if any exceptions escape the split constructor or any member function
+ called on any task, then one of the exceptions (or a copy thereof) propagates
+ on the thread calling this function, but only after all running tasks finish
+ (thus avoiding dangling references); also possibly `std::bad_alloc`.
+
+ After `run` throws on one task, any other task will prompty shut down after it
+ next tests `progress`.
+
+ @pre `n_threads >= 1`
+ */
+template<
+   typename Task,
+   typename Increment
+      = typename std::iterator_traits<typename Task::iterator_type>
+         ::difference_type
+>
+void OneDimensionalReduce(Task &task,
+   unsigned n_threads = std::thread::hardware_concurrency());
+
+enum ReportChoice{ NoReport, Report };
+enum MergeChoice{ NoMerge, Merge };
+
+//! Adapts its base as a task type that checks and reports progress
+/*!
+ Generates a `run` that iterates the task's range, calling operator() on
+ each element, and yielding after each.
+
+ Also, generates a `merge` that yields, and if doMerge, that function calls a
+ one-argument `merge` in the base class.
+
+ If doReport, then these base class functions are expected to return an
+ increment value that is passed to the yielder; also the base must define
+ `report_progress` which accepts the total progress value and is called only on
+ the original task.
+ */
+template<typename Task, typename Increment,
+   MergeChoice doMerge, ReportChoice doReport>
+struct TaskAdapter : Task {
+   using Task::Task;
+   TaskAdapter(
+      TaskAdapter &orig, typename Task::iterator_type iter)
+      : Task{ orig, iter }
+      , is_first{ false }
+   {}
+   void run(Progress<Increment> &progress)
+   {
+      std::for_each(this->begin(), this->end(), [this, &progress](auto &&value){
+         if constexpr (doReport) {
+            auto total =
+               progress((*this)(std::forward<decltype(value)>(value)));
+            if (is_first)
+               this->report_progress(total);
+         }
+         else {
+            (*this)(std::forward<decltype(value)>(value));
```

Comment:
I find such code to be very poorly readable.

Reply:
It's the unfortunate correct way to forward an argument perfectly.